hbase-http/src/test/java/org/apache/hadoop/hbase/http/TestProxyUserSpnegoHttpServer.java

```diff
@@ -0,0 +1,262 @@
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.apache.kerby.kerberos.kerb.KrbException;
+import org.apache.kerby.kerberos.kerb.client.JaasKrbUtil;
+import org.apache.kerby.kerberos.kerb.server.SimpleKdcServer;
+import org.ietf.jgss.GSSCredential;
+import org.ietf.jgss.GSSManager;
+import org.ietf.jgss.GSSName;
+import org.ietf.jgss.Oid;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test class for SPNEGO Proxyuser authentication on the HttpServer. Uses Kerby's MiniKDC and Apache
+ * HttpComponents to verify that the doas= mechanicsm works, and that the proxyuser settings are
+ * observed 
+ */
+@Category({MiscTests.class, SmallTests.class})
+public class TestProxyUserSpnegoHttpServer extends HttpServerFunctionalTest {
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+      HBaseClassTestRule.forClass(TestProxyUserSpnegoHttpServer.class);
+
+  private static final Logger LOG = LoggerFactory.getLogger(TestProxyUserSpnegoHttpServer.class);
+  private static final String KDC_SERVER_HOST = "localhost";
+  private static final String WHEEL_PRINCIPAL = "wheel";
+  private static final String STANDARD_PRINCIPAL = "standard";
+  private static final String DOAS_PRINCIPAL = "doas";
+
+  private static HttpServer server;
+  private static URL baseUrl;
+  private static SimpleKdcServer kdc;
+  private static File infoServerKeytab;
+  private static File wheelKeytab;
+  private static File standardKeytab;
+  private static File doasKeytab;
+
+  @BeforeClass
+  public static void setupServer() throws Exception {
+    Configuration conf = new Configuration();
+    HBaseCommonTestingUtility htu = new HBaseCommonTestingUtility(conf);
+
+    final String serverPrincipal = "HTTP/" + KDC_SERVER_HOST;
+
+    kdc = buildMiniKdc();
+    kdc.start();
+    File keytabDir = new File(htu.getDataTestDir("keytabs").toString());
+    if (keytabDir.exists()) {
+      deleteRecursively(keytabDir);
+    }
+    keytabDir.mkdirs();
+
+    infoServerKeytab = new File(keytabDir, serverPrincipal.replace('/', '_') + ".keytab");
+    wheelKeytab = new File(keytabDir, WHEEL_PRINCIPAL + ".keytab");
+    standardKeytab = new File(keytabDir, STANDARD_PRINCIPAL + ".keytab");
+    doasKeytab = new File(keytabDir, DOAS_PRINCIPAL + ".keytab");
+
+    setupUser(kdc, wheelKeytab, WHEEL_PRINCIPAL);
+    setupUser(kdc, standardKeytab, STANDARD_PRINCIPAL);
+    setupUser(kdc, doasKeytab, DOAS_PRINCIPAL);
+    setupUser(kdc, infoServerKeytab, serverPrincipal);
+
+    buildSpnegoConfiguration(conf, serverPrincipal, infoServerKeytab);
+ 
+    server = createTestServerWithSecurity(conf);
+    server.addPrivilegedServlet("echo", "/echo", EchoServlet.class);
```

Comment:
Awesome that some of the API I changed in HBASE-17115 could be re-used here.

Can you also set `HttpServer.HTTP_SPNEGO_AUTHENTICATION_ADMIN_USERS_KEY`? Right now, we have
* User is not allowed to proxy the user "doas"
* User is allowed to proxy the user "doas"

Elsewhere, I already asked if you could test the scenario:
* User is allowed to proxy the user "doas" but not any other user

I think we're missing one more important scenario:
* User is allowed to proxy the `doAs` user, but the proxied user is not allowed to access HBase UI.

Setup for this test would be something like...
* "Create" another user "unprivileged"
* Add "unprivileged" to `hadoop.proxyuser.wheel.users`
* Execute request as `wheel` with `?doAs=unprivileged` 
* Validate that `unprivileged` is disallowed from accessing the EchoServlet (but the doAs was successful, if we can distinguish between these two authorization failures)

Reply:
Done. 

We have four combinations of privileged/allowed to sudo, I have addedd tests for three of those.

I think that there is no point trying to unsuccessfully sudo to an unprivileged user, as that would fail the same way as trying to sudo unsuccessfully to a privileged user.

Adding the new test uncovered a problem in the test setup, I've fixed that as well.

Reply:
Whew. Almost left you a big comment, but then realized that GH sent me to a copy of the test from the past :)